File: cppmessagerouter/messageserver/message.cpp
/***************************************************************************
                          message.cpp  -  description
                             -------------------
    begin                : Wed Jan 28 2004
    copyright            : (C) 2004 by David Craine
    email                : dcraine@infotether.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "message.h"

Message::Message(){
  this->msgHeaderSet = false;
}

Message::Message(const Message& m) {
  this->msgHeaderSet = m.msgHeaderSet;
  this->from = m.from;
  this->to = m.to;
  this->thread = m.thread;
  this->subject = m.subject;
  this->body = m.body;
}

const Message& Message::operator= (const Message& right) {
  this->msgHeaderSet = right.msgHeaderSet;
  this->from = right.from;
  this->to = right.to;
  this->thread = right.thread;
  this->subject = right.subject;
  this->body = right.body;
}

Message::Message (const string& _from, const string& _to, const string& _thread, const string& _subject, const string& _body) {
  this->msgHeaderSet = false;
  this->from = _from;
  this->to = _to;
  this->thread = _thread;
  this->subject = _subject;
  this->body = _body;
}

Message::~Message(){
}
/** Read property of string from. */
string& Message::getfrom(){
	return from;
}
/** Write property of string from. */
void Message::setfrom( const string& _newVal){
	from = _newVal;
}
/** Read property of string  to. */
string & Message::getto(){
	return to;
}
/** Write property of string  to. */
void Message::setto( const string & _newVal){
	to = _newVal;
}
/** Read property of subject . */
string& Message::getsubject(){
	return subject;
}
/** Write property of subject . */
void Message::setsubject( const string& _newVal){
	 subject = _newVal;
}
/** Read property of string thread. */
string& Message::getthread(){
	return thread;
}
/** Write property of string thread. */
void Message::setthread( const string& _newVal){
	thread = _newVal;
}
/** Read property of string body. */
string& Message::getbody(){
	return body;
}
/** Write property of string body. */
void Message::setbody( const string& _newVal){
	body = _newVal;
}

void Message::setMessageHeader(char *c) {
  msgHeaderSet = true;
  for (int i=0; i<sizeof(msgHeader); i++) {
    msgHeader[i] = c[i];
  }
}


/** No descriptions */
void Message::setMessageHeader(const string &str){
  this->setMessageHeader(str.c_str());
}

char * Message::getMessageHeader() {
  if (!msgHeaderSet) {  //if the message header has not been explicity set then we need to calculate it
    msgHeader[0] = to.length();     //set "to" size
    msgHeader[1] = from.length();
    msgHeader[2] = thread.length();
    msgHeader[3] = subject.length();
    long bodyLength = body.length();
    msgHeader[4] = bodyLength & 4294967295L;
    msgHeader[5] = bodyLength & 16777215L;
    msgHeader[6] = bodyLength & 65535L;
    msgHeader[7] = bodyLength & 255L;
  }
  return msgHeader;

}

void Message::setMessageData(const string &data) {
  msgData =  data;
}

string& Message::getMessageData() {
  if (msgData == "") {  //if the message data has not been explicitly set, then we need to construct it
    msgData = to+from+thread+subject+body;
  }
  return msgData;  
}
  

File: WinMessageRouter/Message.cs
using System;

namespace WinMessageRouter
{
	/// <summary>
	/// Summary description for Message.
	/// </summary>
	public class Message
	{
		private string from="";
		private string to="";
		private string thread="";
		private string body="";
		private string subject="";
		private byte[] msgHeader;
		private string msgData;

		public Message()
		{
		}

		public Message(Message m) 
		{
			this.from = m.from;
			this.to = m.to;
			this.thread = m.thread;
			this.subject = m.subject;
			this.body = m.body;
		}

		public Message(string from, string to, string subject, string thread, string body) 
		{
			this.from = from;
			this.to = to;
			this.subject = subject;
			this.thread = thread;
			this.body = body;
		}


		public string From
		{
			get
			{
				return from;
			}
			set
			{
				from = value;
			}
		}

		public string To
		{
			get
			{
				return to;
			}
			set
			{
				to = value;
			}
		}

		public string Thread
		{
			get
			{
				return thread;
			}
			set
			{
				thread = value;
			}
		}

		public string Body
		{
			get
			{
				return body;
			}
			set
			{
				body = value;
			}
		}

		public string Subject
		{
			get
			{
				return subject;
			}
			set
			{
				subject = value;
			}
		}

		public string MessageData
		{
			get
			{
				if (msgData == null) 
				{
					msgData = to+from+thread+subject+body;
				}
				return msgData;
			}
			set
			{
				msgData = value;
			}
		}

		public object MessageHeader
		{
			get
			{
				msgHeader = new byte[8];
				msgHeader[0] = (byte)to.Length;
				msgHeader[1] = (byte)from.Length;
				msgHeader[2] = (byte)thread.Length;
				msgHeader[3] = (byte)subject.Length;
				long bodyLength = body.Length;
				msgHeader[4] = (byte)((bodyLength & 0xff000000) >> 24);
				msgHeader[5] = (byte)((bodyLength & 0x00ff0000) >> 16);
				msgHeader[6] = (byte)((bodyLength & 0x0000ff00) >> 8);
				msgHeader[7] = (byte)(bodyLength & 0x000000ff);
				return msgHeader;
			}
			set
			{
			}
		}
	}
}
